jetty-core/jetty-siwe/src/main/java/org/eclipse/jetty/security/siwe/EthereumAuthenticator.java

```diff
@@ -0,0 +1,831 @@
+
+                    MultiPartFormData.Parts parts = MultiPartFormData.from(contentSource, request, contentType, config).get();
+                    signature = parts.getFirst("signature").getContentAsString(StandardCharsets.ISO_8859_1);
+                    message = parts.getFirst("message").getContentAsString(StandardCharsets.ISO_8859_1);
+                }
+                default -> throw new ServerAuthException("Unsupported mime type: " + mimeType);
+            };
+
+            // The browser may convert LF to CRLF, EIP4361 specifies to only use LF.
+            message = message.replace("\r\n", "\n");
+            return new SignedMessage(message, signature);
+        }
+        catch (Throwable t)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("error reading SIWE message and signature", t);
+            sendError(request, response, callback, t.getMessage());
+            return null;
+        }
+    }
+
+    protected AuthenticationState handleNonceRequest(Request request, Response response, Callback callback)
+    {
+        String nonce = createNonce(request.getSession(false));
+        response.getHeaders().put(HttpHeader.CONTENT_TYPE, "application/json");
+        ByteBuffer content = BufferUtil.toBuffer("{ \"nonce\": \"" + nonce + "\" }");
+        response.write(true, content, callback);
+        return AuthenticationState.CHALLENGE;
+    }
+
+    private boolean validateSignInWithEthereumToken(SignInWithEthereumToken siwe, SignedMessage signedMessage, Request request, Response response, Callback callback)
+    {
+        Session session = request.getSession(false);
+        if (siwe == null)
+        {
+            sendError(request, response, callback, "failed to parse SIWE message");
+            return false;
+        }
+
+        try
+        {
+            siwe.validate(signedMessage, nonce -> redeemNonce(session, nonce), _domains, _chainIds);
+        }
+        catch (Throwable t)
+        {
+            sendError(request, response, callback, t.getMessage());
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public AuthenticationState validateRequest(Request request, Response response, Callback callback) throws ServerAuthException
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("validateRequest({},{})", request, response);
+
+        String uri = request.getHttpURI().toString();
+        if (uri == null)
+            uri = "/";
+
+        try
+        {
+            Session session = request.getSession(false);
+            if (session == null)
+            {
+                session = request.getSession(true);
+                if (session == null)
+                {
+                    sendError(request, response, callback, "session could not be created");
+                    return AuthenticationState.SEND_FAILURE;
+                }
+            }
+
+            // TODO: verify the sessionID is obtained from a cookie.
```

Comment:
It would be nice not to do a first checkin with TODOs

Reply:
The work to implement this is out of scope of this PR. I had a try and it got a bit complicated.

But I'm not sure this is necessary, this check is commented out for openId, so I included it here as well, but I will remove it.